allow for namespaced mailers

licatajustin opened the pull request:
This PR allows for namespaced mailers.

We ran into a small issue running this with a mounted engine in our Rails app. The `template_path` wasn't accounting for the namespace.

```ruby
template = /path/to/app/views/namespace/user_mailer/mailer_method.html.erb
template_path = template.inspect.split("/").slice(-2, 2).join("/")
# => user_mailer/mailer_method.html.erb
```

I updated the `template_path` variable, to split the `template` at "views", then keep the rest of the path that follows. This accounts to non-namespaced paths, namespaced paths, and paths that have multiple namespaces.

```ruby
template = /path/to/app/views/namespace/user_mailer/mailer_method.html.erb
template_path = template.inspect.split("views")[1][1..-1]
# => namespace/user_mailer/mailer_method.html.erb

template = /path/to/app/views/user_mailer/mailer_method.html.erb
template_path = template.inspect.split("views")[1][1..-1]
# => user_mailer/mailer_method.html.erb

template = /path/to/app/views/namespace/namespace_further/user_mailer/mailer_method.html.erb
template_path = template.inspect.split("views")[1][1..-1]
# => namespace/namespace_further/user_mailer/mailer_method.html.erb
```

Please let me know if anything else is needed! 

Again, much appreciation for maintaining this gem.

billhorsman commented:
That looks good Justin. I'm going to try and get to the bottom of these test failures. I'm pretty sure it's not related to your changes - more likely something else has changed since I last triggered a Travis build.

licatajustin commented:
Thanks @billhorsman! :+1:

billhorsman commented:
Justin. I've updated `.travis.yml` on master which I *think* should fix the Travis build. Would you mind merging master into this branch and updating this PR again please?

licatajustin commented:
It's possible this will break integrations people have, in the case they demodulized their views to account for this. Might be worth a major version bump, since it's technically not backwards compatible. 

What do you think?

billhorsman commented:
Hi Justin. Sorry for the slow response. I'm not very familiar with mounted engines but I wonder whether we can work out what to do by looking more closely at the path?

Can you describe a situation where this would be a breaking change?

licatajustin commented:
Sure, absolutely. And you don't need an engine to test this out, you can test it out by just creating a namespaced mailer! I'll show an example of what I mean.

Let's say this is my mailer:

```ruby
# app/mailers/organizations/user_mailer.rb

module Organization
  class UserMailer
    def send_group_message(group, user)
      ...
    end
  end
end
```

With this mailer, and _without_ the gem, Rails would look for the following template to render.

`app/views/organizations/user_mailer/send_group_message.html.erb`

However, if I were to include the InlineStylesMailer module, it would start looking for this path instead:

`app/views/user_mailer/send_group_message.html.erb`

It doesn't account for the namespace, because we were only slicing the last two elements of the array with `slice(-2, 2)`. This was raising the error, `ActionView::MissingTemplate`. You can see below why it wasn't accounting for the namespace.

```ruby
path = "app/views/organizations/user_mailer/send_group_message.html.erb"
path.split("/").slice(-2, 2)
# > ["user_mailer", "send_group_message.html.erb"]
```

---

I _could_ have just renamed my view and dropped the `organization` namespace, and the emails would have gone through no problem. So if others were to have accounted for the bug like that, then this PR will break their integrations, because now it's going to look for the correct path, _with_ the namespace, if one exists.

Let me know if I need to add clarity anywhere! Thanks!

billhorsman commented:
A perfect explanation, Justin. I thought it was some complicated engine thing but I see your point.

IMHO, it's not really breaking backwards compatibility. This was a bug and this fix means the workaround will no longer work. 

I am just exploring to see whether we can make the code a bit more lenient and look for the template in the namespaced location first and then fallback to the way it used to do it. However, I'm a bit wary of that because while it might ease the upgrade path for some people I think it might confuse other people in the future because it is doing something unexpected. And the people who have done the workaround are probably in a better position to know what to do about it. Hmm.

licatajustin commented:
Yea, I see what you mean. Ultimately, it's your decision, but please let me know if I can help in any way.

>  However, I'm a bit wary of that because while it might ease the upgrade path for some people I think it might confuse other people in the future because it is doing something unexpected.

I'm definitely in agreement here. If I had accounted for this bug in my app, I personally would appreciate a major version bump, so I wouldn't get caught off guard with a minor version bump that breaks something.

licatajustin commented:
Again, much appreciation for this writing and maintaining this gem. Big :+1: for this. I used to get sad dealing with inline CSS in my mailers.

billhorsman commented:
I'll bump this to `1.0.0` before I release the next version to RubyGems. That might be a day or two... I'm going to do a sweep through and see what else I can clean up while I'm at it.

licatajustin commented:
Awesome, let me know if I can help with anything!
